ballerina-shell/modules/shell-core/src/main/java/io/ballerina/shell/invoker/ShellSnippetsInvoker.java

```diff
@@ -0,0 +1,455 @@
+     */
+    protected SingleFileProject getProject(Object context, String templateFile) throws InvokerException {
+        Mustache template = getTemplate(templateFile);
+        try (StringWriter stringWriter = new StringWriter()) {
+            template.execute(stringWriter, context);
+            return getProject(stringWriter.toString());
+        } catch (IOException e) {
+            addErrorDiagnostic("File generation failed: " + e.getMessage());
+            throw new InvokerException(e);
+        }
+    }
+
+    /**
+     * Get the project with the context data.
+     *
+     * @param source Source to use for generating project.
+     * @return Created ballerina project.
+     * @throws InvokerException If file writing failed.
+     */
+    protected SingleFileProject getProject(String source) throws InvokerException {
+        try {
+            File mainBal = writeToFile(source);
+            BuildOptions buildOptions = new BuildOptionsBuilder().offline(true).build();
+            return SingleFileProject.load(mainBal.toPath(), buildOptions);
+        } catch (IOException e) {
+            addErrorDiagnostic("File writing failed: " + e.getMessage());
+            throw new InvokerException(e);
+        }
+    }
+
+    /* Compilation methods */
+
+    /**
+     * Helper method to compile a project and report any errors.
+     * No code generation is done.
+     *
+     * @param project Project to compile.
+     * @return Compilation data.
+     * @throws InvokerException If compilation failed.
+     */
+    protected PackageCompilation compile(Project project) throws InvokerException {
+        try {
+            Module module = project.currentPackage().getDefaultModule();
+            PackageCompilation packageCompilation = project.currentPackage().getCompilation();
+            DiagnosticResult diagnosticResult = packageCompilation.diagnosticResult();
+
+            for (io.ballerina.tools.diagnostics.Diagnostic diagnostic : diagnosticResult.diagnostics()) {
+                DiagnosticSeverity severity = diagnostic.diagnosticInfo().severity();
+                if (severity == DiagnosticSeverity.ERROR) {
+                    addErrorDiagnostic(highlightedDiagnostic(module, diagnostic));
+                    addErrorDiagnostic("Compilation aborted because of errors.");
+                    throw new InvokerException();
+                } else if (severity == DiagnosticSeverity.WARNING) {
+                    addDiagnostic(Diagnostic.warn(highlightedDiagnostic(module, diagnostic)));
+                } else {
+                    addDiagnostic(Diagnostic.debug(diagnostic.message()));
+                }
+            }
+
+            return packageCompilation;
+        } catch (InvokerException e) {
+            throw e;
+        } catch (Exception e) {
+            addErrorDiagnostic("Something went wrong: " + e);
+            throw new InvokerException(e);
+        } catch (Error e) {
```

Comment:
Let us not try to catch Errors. It says in the java doc:

> An Error is a subclass of Throwable that indicates serious problems that a reasonable application should not try to catch. Most such errors are abnormal conditions. The ThreadDeath error, though a "normal" condition, is also a subclass of Error because most applications should not try to catch it. 
https://docs.oracle.com/javase/7/docs/api/java/lang/Error.html